File: AdventOfCode/2025/Day4/Day4.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// Point is a simple structure to hold Grid Coordinates.
// In Python, we used tuples (r, c). In Go, explicit structs are cleaner.
type Point struct {
	r, c int
}

/*
------------------------------------------------------------------------

	HELPER: File Loader

------------------------------------------------------------------------
*/
func loadInput(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

/*
------------------------------------------------------------------------

	PART 1: Static Grid Check

------------------------------------------------------------------------
*/
func solvePart1(grid []string) int {
	rows := len(grid)
	cols := len(grid[0])
	accessibleCount := 0

	// Define the 8 directions (neighbor offsets)
	// We use a slice of Points to store the offsets
	moves := []Point{
		{-1, -1}, {-1, 0}, {-1, 1}, // Top Row
		{0, -1}, {0, 1}, // Middle Row
		{1, -1}, {1, 0}, {1, 1}, // Bottom Row
	}

	// Loop through every Row and Column
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {

			// GO CONCEPT: Accessing Characters
			// grid[r][c] gives us a 'byte' (ASCII value).
			// We verify if it is equal to the byte for '@'.
			if grid[r][c] != '@' {
				continue
			}

			neighborPaperCount := 0

			// Iterate through the 8 directions
			for _, move := range moves {
				newR := r + move.r
				newC := c + move.c

				// BOUNDARY CHECK
				// Go logic is identical to Python here.
				// Check if we are outside the valid index range.
				if newR < 0 || newR >= rows || newC < 0 || newC >= cols {
					continue
				}

				// Check neighbor content
				if grid[newR][newC] == '@' {
					neighborPaperCount++
				}
			}

			// Apply the Rule: Less than 4 neighbors
			if neighborPaperCount < 4 {
				accessibleCount++
			}
		}
	}

	return accessibleCount
}

/*
------------------------------------------------------------------------

	PART 2: Dynamic Simulation

------------------------------------------------------------------------
*/
func solvePart2(input []string) int {
	/*
	   GO CONCEPT: Mutable Grids
	   In Go, 'string' types are IMMUTABLE (read-only), just like Python.
	   To modify the grid (change '@' to '.'), we must convert it into
	   a "Slice of Slices of Runes" ([][]rune).

	   - A 'rune' is Go's name for a Unicode character.
	   - We allocate the outer slice, then loop to convert each string row.
	*/
	rows := len(input)
	cols := len(input[0])

	// Create the grid container
	grid := make([][]rune, rows)

	for i, rowStr := range input {
		// Convert string "@@.." to slice of runes ['@', '@', '.', '.']
		grid[i] = []rune(rowStr)
	}

	totalRemoved := 0

	// Define moves again (same as Part 1)
	moves := []Point{
		{-1, -1}, {-1, 0}, {-1, 1},
		{0, -1}, {0, 1},
		{1, -1}, {1, 0}, {1, 1},
	}

	// START SIMULATION LOOP
	// Go doesn't have 'while'. We use 'for' without conditions for infinite loops.
	for {
		// 'changes' acts as our buffer.
		// It holds the coordinates (Points) of items to remove this round.
		// We initialize it with capacity 0.
		var changes []Point

		// --- SCAN PHASE ---
		for r := 0; r < rows; r++ {
			for c := 0; c < cols; c++ {

				// Only check existing paper
				if grid[r][c] != '@' {
					continue
				}

				neighborCount := 0

				// Check 8 neighbors
				for _, move := range moves {
					newR := r + move.r
					newC := c + move.c

					// Bounds Check
					if newR < 0 || newR >= rows || newC < 0 || newC >= cols {
						continue
					}

					// Check content
					if grid[newR][newC] == '@' {
						neighborCount++
					}
				}

				// Check Rule
				if neighborCount < 4 {
					// Add this coordinate to our batch list
					// append() is how we add to lists in Go
					changes = append(changes, Point{r, c})
				}
			}
		}

		// --- UPDATE PHASE ---

		// Stop Condition: If changes slice is empty, we are done.
		if len(changes) == 0 {
			break
		}

		// Add to total
		totalRemoved += len(changes)

		// Apply the batch updates
		// We loop through our 'changes' slice
		for _, p := range changes {
			// p is a Point struct {r, c}
			// We can safely modify the grid now because it is [][]rune
			grid[p.r][p.c] = '.'
		}
	}

	return totalRemoved
}

/*
------------------------------------------------------------------------

	MAIN ENTRY POINT

------------------------------------------------------------------------
*/
func main() {
	// 1. SAMPLE DATA
	sampleData := []string{
		"..@@.@@@@.",
		"@@@.@.@.@@",
		"@@@@@.@.@@",
		"@.@@@@..@.",
		"@@.@@@@.@@",
		".@@@@@@@.@",
		".@.@.@.@@@",
		"@.@@@.@@@@",
		".@@@@@@@@.",
		"@.@.@@@.@.",
	}

	fmt.Println("--- SAMPLE DATA ---")
	fmt.Printf("Part 1: %d\n", solvePart1(sampleData))
	fmt.Printf("Part 2: %d\n", solvePart2(sampleData))
	fmt.Println()

	// 2. REAL DATA
	filename := "day4-data.txt"
	realData, err := loadInput(filename)

	if err != nil {
		fmt.Printf("Error reading %s: %v\n", filename, err)
	} else {
		fmt.Println("--- REAL DATA ---")
		fmt.Printf("Part 1: %d\n", solvePart1(realData))
		fmt.Printf("Part 2: %d\n", solvePart2(realData))
	}
}

File: AdventOfCode/2025/Day4/Day4.py
"""
------------------------------------------------------------------------
Challenge: Day 4 - Printing Department (Part 1)
------------------------------------------------------------------------
Author: Jac-Now

Context:
    We need to determine which paper rolls ('@') are accessible to forklifts.
    A paper roll is accessible if it is NOT crowded.

Mechanics:
    - The warehouse is a 2D grid of characters.
    - '@' represents a paper roll. '.' represents empty space.
    - A roll is accessible if it has FEWER than 4 neighbors (up/down/left/right/diagonal).

Goal:
    - Scan every cell in the grid.
    - For every '@', count its '@' neighbors.
    - If neighbors < 4, it counts towards the answer.
------------------------------------------------------------------------
"""
import os

def load_input(filename):
    """
    Reads the input file into a list of strings.
    Each string represents one ROW in the grid.
    """
    script_dir = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(script_dir, filename)
    
    if not os.path.exists(file_path):
        return []
        
    with open(file_path, 'r') as f:
        # .splitlines() removes the \n characters from the end of lines
        return f.read().splitlines()

def count_accessible_paper(grid):
    """
    Main logic to solve the puzzle.
    Args:
        grid (list of str): The 2D map of the warehouse.
    Returns:
        int: The total number of accessible paper rolls.
    """
    # 1. Determine Grid Dimensions
    # 'rows' is how many lines of text we have (Height)
    rows = len(grid)
    # 'cols' is how many characters are in the first line (Width)
    cols = len(grid[0])
    
    accessible_count = 0
    
    # 2. Define Relative Moves
    # These represent the 8 directions around a specific point (r, c).
    # Format: (change_in_row, change_in_col)
    moves = [
        (-1, -1), (-1, 0), (-1, 1), # Top-Left, Top, Top-Right
        (0, -1),           (0, 1),  # Left,      Right
        (1, -1),  (1, 0),  (1, 1)   # Bot-Left, Bot, Bot-Right
    ]

    # 3. Iterate through every cell in the grid
    # We use nested loops: Outer loop for rows, Inner loop for columns.
    for r in range(rows):
        for c in range(cols):
            
            # OPTIMIZATION:
            # If the current cell is empty ('.'), we don't need to check it.
            # We only care about paper rolls ('@').
            if grid[r][c] != '@':
                continue
                
            # 4. Count the Neighbors for this specific paper roll
            neighbor_paper_count = 0
            
            # Check all 8 directions defined in 'moves'
            for dr, dc in moves:
                # Calculate the neighbor's coordinate
                new_r = r + dr
                new_c = c + dc
                
                # SAFETY CHECK (CRITICAL):
                # We must ensure the neighbor coordinate is actually inside the grid.
                # If new_r is -1, we are looking off the top edge.
                # If new_r is equal to 'rows', we are looking off the bottom edge.
                if new_r < 0 or new_r >= rows or new_c < 0 or new_c >= cols:
                    continue # Skip this neighbor, it doesn't exist (it's void)
                
                # If the neighbor is valid, check if it contains paper
                if grid[new_r][new_c] == '@':
                    neighbor_paper_count += 1
            
            # 5. Apply the Rule
            # "fewer than 4 rolls of paper in the 8 adjacent positions"
            if neighbor_paper_count < 4:
                accessible_count += 1
                
    return accessible_count

"""
------------------------------------------------------------------------
Challenge: Day 4 - Printing Department (Part 2)
------------------------------------------------------------------------
Context:
    The problem shifts from a static check to a dynamic simulation.
    Once a paper roll is accessible, it is REMOVED from the warehouse.
    Removing a roll creates empty space, which might make OTHER rolls 
    accessible in the next "round".

Mechanics:
    - Cellular Automata / Simulation Loop.
    - We must identify ALL rolls that are accessible in the current configuration.
    - We remove them ALL AT ONCE (Batch Update).
    - We repeat this process until no more rolls can be removed.

Goal:
    - Count the total number of paper rolls removed throughout the entire simulation.
------------------------------------------------------------------------
"""

def count_accessible_paper_part2(raw_grid):
    """
    Simulates the repetitive removal of paper rolls.
    """
    # ----------------------------------------------------------------
    # SETUP PHASE
    # ----------------------------------------------------------------
    # Python strings (e.g., "@@..") are immutable, meaning we cannot change 
    # individual characters. We must convert the grid into a List of Lists 
    # to allow us to update the grid (grid[r][c] = '.') later.
    grid = [list(row) for row in raw_grid]
    
    accessible_count = 0
    
    # Define Relative Moves (Vector Arithmetic)
    # These represent the 8 directions around a specific point (r, c).
    # Format: (change_in_row, change_in_col)
    moves = [
        (-1, -1), (-1, 0), (-1, 1), # Top-Left, Top, Top-Right
        (0, -1),           (0, 1),  # Left,      Right
        (1, -1),  (1, 0),  (1, 1)   # Bot-Left, Bot, Bot-Right
    ]

    # ----------------------------------------------------------------
    # SIMULATION LOOP
    # ----------------------------------------------------------------
    # We use 'while True' because we don't know how many rounds it will take.
    # We will break out manually when no changes occur.
    while True:
        
        # 'changes' acts as a temporary buffer.
        # We cannot remove items immediately because that would change the 
        # neighbor counts for other cells in the *same* round.
        # This enforces the "Simultaneous Update" rule.
        changes = []
        
        rows = len(grid)
        cols = len(grid[0])

        # --- SCAN PHASE ---
        # Iterate through every cell to find candidates for removal
        for r in range(rows):
            for c in range(cols):
                
                # OPTIMIZATION:
                # Skip empty spots. We only care about existing paper rolls.
                if grid[r][c] != '@':
                    continue
                    
                # Count Neighbors
                neighbor_paper_count = 0
                
                # Check all 8 directions
                for dr, dc in moves:
                    new_r = r + dr
                    new_c = c + dc
                    
                    # SAFETY CHECK (Boundary Validation):
                    # Ensure we don't try to look outside the grid limits.
                    # 1. new_r must be >= 0 and < total rows
                    # 2. new_c must be >= 0 and < total cols
                    if new_r < 0 or new_r >= rows or new_c < 0 or new_c >= cols:
                        continue # It's off the map (void), so it's not paper.
                    
                    # If valid neighbor, check for paper
                    if grid[new_r][new_c] == '@':
                        neighbor_paper_count += 1
                
                # APPLY THE RULE:
                # "Fewer than 4 rolls of paper in the 8 adjacent positions"
                if neighbor_paper_count < 4:
                    accessible_count += 1
                    # Mark for removal, but don't remove yet!
                    changes.append((r, c))
        
        # --- UPDATE PHASE ---
        
        # STOP CONDITION:
        # If the 'changes' list is empty, it means the grid has stabilized.
        # No more paper can be reached. We are done.
        if changes == []:
            break
            
        # Process the Batch Updates
        for r, c in changes:
            # Change '@' to '.' (Empty Space)
            # This will affect the neighbor counts in the NEXT round.
            grid[r][c] = '.'
            
    # Return the total accumulated count after the simulation stops
    return accessible_count


# ==========================================
# Main Execution
# ==========================================
if __name__ == "__main__":
    
    # 1. Run Example Data
    # -------------------
    sample_data = [
        "..@@.@@@@.",
        "@@@.@.@.@@",
        "@@@@@.@.@@",
        "@.@@@@..@.",
        "@@.@@@@.@@",
        ".@@@@@@@.@",
        ".@.@.@.@@@",
        "@.@@@.@@@@",
        ".@@@@@@@@.",
        "@.@.@@@.@."
    ]
    
    print("--- Example Data ---")
    print(f"Result: {count_accessible_paper(sample_data)}")
    

    # 2. Run Real Data
    # ----------------
    real_data = load_input('day4-data.txt')
    if real_data:
        print("\n--- Real Data ---")
        print(f"Part 1 Answer: {count_accessible_paper(real_data)}")
    else:
        print("\nReal data file not found (day4-data.txt). Skipping.")

     # 3. Part 2 Execution
    # ---------------------
    print("\n--- Part 2 Example Data ---")
    print(f"Result: {count_accessible_paper_part2(sample_data)}")
    if real_data:
        print("\n--- Part 2 Real Data ---")
        print(f"Part 2 Answer: {count_accessible_paper_part2(real_data)}")


